frontend/apps/reader/modules/readertoc.lua

```diff
@@ -60,10 +60,13 @@ end
 
 function ReaderToc:onPageUpdate(pageno)
     self.pageno = pageno
-    if UIManager.FULL_REFRESH_COUNT == -1 then
+    if UIManager.FULL_REFRESH_COUNT == -1 or G_reader_settings:isTrue("refresh_on_chapter_boundaries") then
         if self:isChapterEnd(pageno, 0) then
             self.chapter_refresh = true
-        elseif self:isChapterBegin(pageno, 0) and self.chapter_refresh then
+        elseif self:isChapterStart(pageno, 0) and self.chapter_refresh then
+            UIManager:setDirty(nil, "full")
+            self.chapter_refresh = true
```

Comment:
shouldn't it be `false`? Otherwise it's not needed, since the requirement to enter this if is `self.chapter_refresh == true`

Reply:
It's indeed redundant, but I kept it to keep the ladder "symmetric" in order to make the logic more obvious ;).

Reply:
I'm half of a mind to zap that heuristic and just flash on start/second no matter the pagination direction...

Reply:
Mildly intrigued that luacheck doesn't catch that as a dead store, by the way...